src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -430,35 +453,37 @@ matches deps TypeClassDictionaryInScope{..} tys = do
-    typeHeadsAreEqual (Skolem _ s1 _ _)    (Skolem _ s2 _ _)    | s1 == s2 = (True, M.empty)
-    typeHeadsAreEqual t                    (TypeVar v)                     = (True, M.singleton v [t])
-    typeHeadsAreEqual (TypeConstructor c1) (TypeConstructor c2) | c1 == c2 = (True, M.empty)
-    typeHeadsAreEqual (TypeLevelString s1) (TypeLevelString s2) | s1 == s2 = (True, M.empty)
+    typeHeadsAreEqual (TUnknown u1)        (TUnknown u2)        | u1 == u2 = (Match (), M.empty)
+    typeHeadsAreEqual (Skolem _ s1 _ _)    (Skolem _ s2 _ _)    | s1 == s2 = (Match (), M.empty)
+    typeHeadsAreEqual t                    (TypeVar v)                     = (Match (), M.singleton v [t])
+    typeHeadsAreEqual (TypeConstructor c1) (TypeConstructor c2) | c1 == c2 = (Match (), M.empty)
+    typeHeadsAreEqual (TypeLevelString s1) (TypeLevelString s2) | s1 == s2 = (Match (), M.empty)
     typeHeadsAreEqual (TypeApp h1 t1)      (TypeApp h2 t2)                 =
       both (typeHeadsAreEqual h1 h2) (typeHeadsAreEqual t1 t2)
-    typeHeadsAreEqual REmpty REmpty = (True, M.empty)
+    typeHeadsAreEqual REmpty REmpty = (Match (), M.empty)
     typeHeadsAreEqual r1@RCons{} r2@RCons{} =
         foldr both (uncurry go rest) common
       where
         (common, rest) = alignRowsWith typeHeadsAreEqual r1 r2
 
-        go :: ([(Label, Type)], Type) -> ([(Label, Type)], Type) -> (Bool, Matching [Type])
+        go :: ([(Label, Type)], Type) -> ([(Label, Type)], Type) -> (Matched (), Matching [Type])
         go (l,  KindedType t1 _)  (r,  t2)                            = go (l, t1) (r, t2)
         go (l,  t1)               (r,  KindedType t2 _)               = go (l, t1) (r, t2)
-        go ([], REmpty)           ([], REmpty)                        = (True, M.empty)
-        go ([], TUnknown u1)      ([], TUnknown u2)      | u1 == u2   = (True, M.empty)
-        go ([], TypeVar v1)       ([], TypeVar v2)       | v1 == v2   = (True, M.empty)
-        go ([], Skolem _ sk1 _ _) ([], Skolem _ sk2 _ _) | sk1 == sk2 = (True, M.empty)
-        go (sd, r)                ([], TypeVar v)                     = (True, M.singleton v [rowFromList (sd, r)])
-        go _ _                                                        = (False, M.empty)
-    typeHeadsAreEqual _ _ = (False, M.empty)
-
-    both :: (Bool, Matching [Type]) -> (Bool, Matching [Type]) -> (Bool, Matching [Type])
-    both (b1, m1) (b2, m2) = (b1 && b2, M.unionWith (++) m1 m2)
+        go ([], REmpty)           ([], REmpty)                        = (Match (), M.empty)
+        go ([], TUnknown u1)      ([], TUnknown u2)      | u1 == u2   = (Match (), M.empty)
+        go ([], TypeVar v1)       ([], TypeVar v2)       | v1 == v2   = (Match (), M.empty)
+        go ([], Skolem _ sk1 _ _) ([], Skolem _ sk2 _ _) | sk1 == sk2 = (Match (), M.empty)
+        go (sd, r)                ([], TypeVar v)                     = (Match (), M.singleton v [rowFromList (sd, r)])
```

Comment:
I've been trying for a little while, but I'm not sure how to get this situation to happen.  Any ideas?